module/executiondatasync/optimistic_syncing/core.go

```diff
@@ -2,18 +2,266 @@ package pipeline
+	header *flow.Header,
+	execDataRequester requester.ExecutionDataRequester,
+	txResultErrMsgsRequester tx_error_messages.TransactionResultErrorMessageRequester,
+	txResultErrMsgsRequestTimeout time.Duration,
+	persistentRegisters storage.RegisterIndex,
+	persistentEvents storage.Events,
+	persistentCollections storage.Collections,
+	persistentTransactions storage.Transactions,
+	persistentResults storage.LightTransactionResults,
+	persistentTxResultErrMsg storage.TransactionResultErrorMessages,
+	protocolDB storage.DB,
+) *CoreImpl {
+	coreLogger := logger.With().
+		Str("component", "execution_data_core").
+		Str("execution_result_id", executionResult.ID().String()).
+		Str("block_id", executionResult.BlockID.String()).
+		Uint64("height", header.Height).
+		Logger()
+
+	inmemRegisters := unsynchronized.NewRegisters(header.Height)
+	inmemEvents := unsynchronized.NewEvents()
+	inmemCollections := unsynchronized.NewCollections()
+	inmemTransactions := unsynchronized.NewTransactions()
+	inmemResults := unsynchronized.NewLightTransactionResults()
+	inmemTxResultErrMsgs := unsynchronized.NewTransactionResultErrorMessages()
+
+	indexerComponent := indexer.NewInMemoryIndexer(
+		coreLogger,
+		inmemRegisters,
+		inmemEvents,
+		inmemCollections,
+		inmemTransactions,
+		inmemResults,
+		inmemTxResultErrMsgs,
+		executionResult,
+		header,
+	)
+
+	persisterComponent := indexer.NewPersister(
+		coreLogger,
+		inmemRegisters,
+		inmemEvents,
+		inmemCollections,
+		inmemTransactions,
+		inmemResults,
+		inmemTxResultErrMsgs,
+		persistentRegisters,
+		persistentEvents,
+		persistentCollections,
+		persistentTransactions,
+		persistentResults,
+		persistentTxResultErrMsg,
+		protocolDB,
+		executionResult,
+		header,
+	)
+
+	return &CoreImpl{
+		log: coreLogger,
+		processor: &executionDataProcessor{
+			execDataRequester:             execDataRequester,
+			txResultErrMsgsRequester:      txResultErrMsgsRequester,
+			txResultErrMsgsRequestTimeout: txResultErrMsgsRequestTimeout,
+			indexer:                       indexerComponent,
+			persister:                     persisterComponent,
+			inmemRegisters:                inmemRegisters,
+			inmemEvents:                   inmemEvents,
+			inmemCollections:              inmemCollections,
+			inmemTransactions:             inmemTransactions,
+			inmemResults:                  inmemResults,
+			inmemTxResultErrMsgs:          inmemTxResultErrMsgs,
+		},
+		executionResult: executionResult,
+		header:          header,
+	}
+}
+
+// Download downloads execution data and transaction results error for the block
+// Expected errors:
+// - context.Canceled: if the provided context was canceled before completion
+// - context.DeadlineExceeded: if the provided context was canceled due to its deadline reached
+//
+// No other errors are expected during normal operations
+func (c *CoreImpl) Download(ctx context.Context) error {
+	c.log.Debug().Msg("downloading execution data")
+
+	g, gCtx := errgroup.WithContext(ctx)
```

Comment:
using `WithContext` here means if either fail, the context is canceled causing both to fail. This is generally OK since neither requester should fail except if there is an exception. We may need to check for context cancellation or deadline exceeded and handle appropriately. check the implementations.